README.md
# RBush

RBush is a high-performance Dart library for 2D **spatial indexing** of points and rectangles.
It's based on an optimized **R-tree** data structure with **bulk insertion** support.

*Spatial index* is a special data structure for points and rectangles
that allows you to perform queries like "all items within this bounding box" very efficiently
(e.g. hundreds of times faster than looping over all items).
It's most commonly used in maps and data visualizations.

## Usage

```dart
// Create a tree with up to 16 elements in a leaf.
final tree = RBush(16);

// Bulk load elements (empty in this case, so here it's a no-op).
tree.load(<RBushElement>[]);

// Insert a single element.
tree.insert(RBushElement(
  minX: 10, minY: 10,
  maxX: 20, maxY: 30,
  data: 'sample data'
));

// Find the element we've inserted.
final List<RBushElement> found = tree.search(
    RBushBox(minX: 5, minY: 5, maxX: 25, maxY: 25));

// Remove all elements from the tree.
tree.clear();
```

An optional argument to `RBush` defines the maximum number of entries in a tree node.
`9` (used by default) is a reasonable choice for most applications.
Higher value means faster insertion and slower search, and vice versa.

To store items of a different type, extend from (or instantiate) `RBushBase<T>`.
For example:

```dart
class MyItem {
  final String id;
  final LatLng location;
  
  const MyItem(this.id, this.location);
}

final tree = RBushBase<MyItem>(
  maxEntries: 4,
  toBBox: (item) => RBushBox(
    minX: item.location.longitude,
    maxX: item.location.longitude,
    minY: item.location.latitude,
    maxY: item.location.latitude,
  ),
  getMinX: (item) => item.location.longitude,
  getMinY: (item) => item.location.latitude,
);
```

### K Nearest Neighbours

The `RBushBase` class also includes a `knn()` method for the nearest neighbours
search. This is especially useful when using the r-tree to store point features,
like in the example above.

Note that for larger geographical areas the distance would be wrong, since the
class uses pythagorean distances (`dx² + dy²`), not Haversine or great circle.

## Tiny Queue and Quick Select

This package also includes ported fast versions of a priority queue and
a selection algorithm. These are used internally by the r-tree, but might
be useful for you as well.

## Upstream

This library is a straight-up port of several JavaScript libraries written
by Vladimir Agafonkin:

* [RBush 3.0.1](https://github.com/mourner/rbush)
* [RBush-knn 3.0.1](https://github.com/mourner/rbush-knn)
* [QuickSelect 2.0.0](https://github.com/mourner/quickselect)
* [TinyQueue 2.0.3](https://github.com/mourner/tinyqueue)

All of these are published under MIT or ISC licenses.

CHANGELOG.md
## 1.1.1

* Better node children sorting (thanks @brand17).
* Upgraded dependencies and enabled Dart 3.

## 1.1.0

* `RBushElement` is now templated and requires a non-null data.
* Added `RBushDirect<T>` for skipping bounding boxes management.

## 1.0.0

* Initial version.


test/rbush.dart
import 'package:test/test.dart';
import 'package:rbush/rbush.dart';

List<RBushElement> someData(int n) {
  final List<RBushElement> data = [];
  for (double i = 0; i < n; i++) {
    data.add(RBushElement(minX: i, minY: i, maxX: i, maxY: i, data: i));
  }
  return data;
}

RBushElement listToBBox(List<double> list) => RBushElement.fromList(list, 0);

final data = <List<double>>[
  [0, 0, 0, 0], [10, 10, 10, 10], [20, 20, 20, 20], [25, 0, 25, 0], [35, 10, 35, 10],
  [45, 20, 45, 20], [0, 25, 0, 25], [10, 35, 10, 35], [20, 45, 20, 45], [25, 25, 25, 25],
  [35, 35, 35, 35], [45, 45, 45, 45], [50, 0, 50, 0], [60, 10, 60, 10], [70, 20, 70, 20],
  [75, 0, 75, 0], [85, 10, 85, 10], [95, 20, 95, 20], [50, 25, 50, 25], [60, 35, 60, 35],
  [70, 45, 70, 45], [75, 25, 75, 25], [85, 35, 85, 35], [95, 45, 95, 45], [0, 50, 0, 50],
  [10, 60, 10, 60], [20, 70, 20, 70], [25, 50, 25, 50], [35, 60, 35, 60], [45, 70, 45, 70],
  [0, 75, 0, 75], [10, 85, 10, 85], [20, 95, 20, 95], [25, 75, 25, 75], [35, 85, 35, 85],
  [45, 95, 45, 95], [50, 50, 50, 50], [60, 60, 60, 60], [70, 70, 70, 70], [75, 50, 75, 50],
  [85, 60, 85, 60], [95, 70, 95, 70], [50, 75, 50, 75], [60, 85, 60, 85], [70, 95, 70, 95],
  [75, 75, 75, 75], [85, 85, 85, 85], [95, 95, 95, 95],
].map((list) => listToBBox(list)).toList();

final emptyData = <List<double>>[
  [double.negativeInfinity, double.negativeInfinity, double.infinity, double.infinity],
  [double.negativeInfinity, double.negativeInfinity, double.infinity, double.infinity],
  [double.negativeInfinity, double.negativeInfinity, double.infinity, double.infinity],
  [double.negativeInfinity, double.negativeInfinity, double.infinity, double.infinity],
  [double.negativeInfinity, double.negativeInfinity, double.infinity, double.infinity],
  [double.negativeInfinity, double.negativeInfinity, double.infinity, double.infinity],
].map((list) => listToBBox(list));

class MyItem {
  final double minLng;
  final double minLat;
  final double maxLng;
  final double maxLat;

  const MyItem(
      {required this.minLng,
      required this.minLat,
      required this.maxLng,
      required this.maxLat});

  @override
  bool operator ==(Object other) {
    if (other is! MyItem) return false;
    return minLng == other.minLng && minLat == other.minLat
        && maxLng == other.maxLng && maxLat == other.maxLat;
  }
}

void main() {
  test('allows custom formats by overriding some methods', () {
    final tree = RBushBase<MyItem>(
      toBBox: (item) => RBushBox(
        minX: item.minLng,
        minY: item.minLat,
        maxX: item.maxLng,
        maxY: item.maxLat,
      ),
      getMinX: (item) => item.minLng,
      getMinY: (item) => item.minLat,
    );
    expect(tree.toBBox(MyItem(minLng: 1, minLat: 2, maxLng: 3, maxLat: 4)),
        equals(RBushBox(minX: 1, minY: 2, maxX: 3, maxY: 4)));
  });

  test('constructor uses 9 max entries by default', () {
    final tree = RBush().load(someData(9));
    expect(tree.data.height, equals(1));

    final tree2 = RBush().load(someData(10));
    expect(tree2.data.height, equals(2));
  });

  test('#toBBox, #compareMinX, #compareMinY can be overriden to allow custom data structures', () {
    final tree = RBushBase<MyItem>(
      maxEntries: 4,
      toBBox: (item) => RBushBox(
        minX: item.minLng,
        minY: item.minLat,
        maxX: item.maxLng,
        maxY: item.maxLat,
      ),
      getMinX: (item) => item.minLng,
      getMinY: (item) => item.minLat,
    );

    const data = [
      MyItem(minLng: -115, minLat: 45, maxLng: -105, maxLat: 55),
      MyItem(minLng: 105, minLat: 45, maxLng: 115, maxLat: 55),
      MyItem(minLng: 105, minLat: -55, maxLng: 115, maxLat: -45),
      MyItem(minLng: -115, minLat: -55, maxLng: -105, maxLat: -45),
    ];

    tree.load(data);

    expect(
      tree.search(RBushBox(minX: -180, minY: -90, maxX: 180, maxY: 90)),
      unorderedEquals([
        MyItem(minLng: -115, minLat: 45, maxLng: -105, maxLat: 55),
        MyItem(minLng: 105, minLat: 45, maxLng: 115, maxLat: 55),
        MyItem(minLng: 105, minLat: -55, maxLng: 115, maxLat: -45),
        MyItem(minLng: -115, minLat: -55, maxLng: -105, maxLat: -45),
      ]),
    );

    expect(
      tree.search(RBushBox(minX: -180, minY: -90, maxX: 0, maxY: 90)),
      unorderedEquals([
        MyItem(minLng: -115, minLat: 45, maxLng: -105, maxLat: 55),
        MyItem(minLng: -115, minLat: -55, maxLng: -105, maxLat: -45),
      ]),
    );

    expect(
      tree.search(RBushBox(minX: 0, minY: -90, maxX: 180, maxY: 90)),
      unorderedEquals([
        MyItem(minLng: 105, minLat: 45, maxLng: 115, maxLat: 55),
        MyItem(minLng: 105, minLat: -55, maxLng: 115, maxLat: -45),
      ]),
    );

    expect(
      tree.search(RBushBox(minX: -180, minY: 0, maxX: 180, maxY: 90)),
      unorderedEquals([
        MyItem(minLng: -115, minLat: 45, maxLng: -105, maxLat: 55),
        MyItem(minLng: 105, minLat: 45, maxLng: 115, maxLat: 55),
      ]),
    );

    expect(
      tree.search(RBushBox(minX: -180, minY: -90, maxX: 180, maxY: 0)),
      unorderedEquals([
        MyItem(minLng: 105, minLat: -55, maxLng: 115, maxLat: -45),
        MyItem(minLng: -115, minLat: -55, maxLng: -105, maxLat: -45),
      ]),
    );
  });

  test('RBushDirect returns custom data types', () {
    final tree = RBushDirect<int>(4);

    final data = [
      RBushElement.fromList([-115, 45, -105, 55], 1),
      RBushElement.fromList([105, 45, 115, 55], 2),
      RBushElement.fromList([105, -55, 115, -45], 3),
    ];

    tree.load(data);
    tree.insert(RBushBox.fromList([-115, -55, -105, -45]), 4);

    expect(tree.all(), unorderedEquals([1, 2, 3, 4]));

    expect(
      tree.search(RBushBox(minX: -180, minY: -90, maxX: 180, maxY: 90)),
      unorderedEquals([1, 2, 3, 4]),
    );

    expect(
      tree.search(RBushBox(minX: -180, minY: -90, maxX: 0, maxY: 90)),
      unorderedEquals([1, 4]),
    );

    expect(
      tree.collides(RBushBox(minX: -180, minY: -90, maxX: 0, maxY: 90)),
      isTrue,
    );

    tree.remove(3);
    expect(tree.all(), unorderedEquals([1, 2, 4]));

    expect(() => tree.insert(RBushBox.fromList([0, 0, 1, 1]), 1), throwsStateError,
        reason: 'no duplicates in RBushDirect');

    tree.clear();
    expect(tree.all(), isEmpty);
  });

  test('#load bulk-loads the given data given max node entries and forms a proper search tree', () {
    final tree = RBush(4).load(data);
    expect(tree.all(), unorderedEquals(data));
  });

  test('#load uses standard insertion when given a low number of items', () {
    final tree = RBush(8).load(data).load(data.sublist(0, 3));
    final tree2 = RBush(8).load(data);

    tree2.insert(data[0]);
    tree2.insert(data[1]);
    tree2.insert(data[2]);

    expect(tree.data, equals(tree2.data));
  });

  test('#load does nothing if loading empty data', () {
    final tree = RBush().load([]);
    expect(tree.data, equals(RBush().data));
  });

  test('#load handles the insertion of maxEntries + 2 empty bboxes', () {
    final tree = RBush(4).load(emptyData);
    expect(tree.data.height, equals(2));
    expect(tree.all(), unorderedEquals(emptyData));
  });

  test('#insert handles the insertion of maxEntries + 2 empty bboxes', () {
    final tree = RBush(4);
    for (final element in emptyData) { tree.insert(element); }

    expect(tree.data.height, equals(2), reason: 'height');
    expect(tree.all(), unorderedEquals(emptyData), reason: 'all');
    expect(tree.data.children[0].childrenLength, equals(4), reason: 'children 0');
    expect(tree.data.children[1].childrenLength, equals(2), reason: 'children 1');
  });

  test('#load properly splits tree root when merging trees of the same height', () {
    final tree = RBush(4).load(data).load(data);
    expect(tree.data.height, equals(4));
    expect(tree.all(), unorderedEquals(data + data));
  });

  test('#load properly merges data of smaller or bigger tree heights', () {
    final smaller = someData(10);
    final tree1 = RBush(4).load(data).load(smaller);
    final tree2 = RBush(4).load(smaller).load(data);

    expect(tree1.data.height, equals(tree2.data.height));
    expect(tree1.all(), unorderedEquals(data + smaller));
    expect(tree2.all(), unorderedEquals(data + smaller));
  });

  test('#search finds matching points in the tree given a bbox', () {
    final tree = RBush(4).load(data);
    final result = tree.search(RBushBox(minX: 40, minY: 20, maxX: 80, maxY: 70));

    expect(result, unorderedEquals(<List<double>>[
      [70,20,70,20],[75,25,75,25],[45,45,45,45],[50,50,50,50],[60,60,60,60],[70,70,70,70],
      [45,20,45,20],[45,70,45,70],[75,50,75,50],[50,25,50,25],[60,35,60,35],[70,45,70,45],
    ].map((list) => listToBBox(list))));
  });

  test('#collides returns true when search finds matching points', () {
    final tree = RBush(4).load(data);
    final result = tree.collides(RBushBox(minX: 40, minY: 20, maxX: 80, maxY: 70));

    expect(result, isTrue);
  });

  test('#search returns an empty array if nothing found', () {
    final result = RBush(4).load(data).search(RBushBox(minX: 200, minY: 200, maxX: 200, maxY: 200));
    expect(result, equals([]));
  });

  test('#collides returns false if nothing found', () {
    final result = RBush(4).load(data).collides(RBushBox(minX: 200, minY: 200, maxX: 200, maxY: 200));
    expect(result, isFalse);
  });

  test('#all returns all points in the tree', () {
    final tree = RBush(4).load(data);
    final result = tree.all();

    expect(result, unorderedEquals(data));
    expect(tree.search(RBushBox(minX: 0, minY: 0, maxX: 100, maxY: 100)), unorderedEquals(data));
  });

  test('#insert adds an item to an existing tree correctly', () {
    final items = <List<double>>[
      [0, 0, 0, 0],
      [1, 1, 1, 1],
      [2, 2, 2, 2],
      [3, 3, 3, 3],
      [1, 1, 2, 2],
    ].map((list) => listToBBox(list)).toList();

    final tree = RBush(4).load(items.sublist(0, 3));

    tree.insert(items[3]);
    expect(tree.data.height, equals(1), reason: 'height 3');
    expect(tree.all(), unorderedEquals(items.sublist(0, 4)), reason: 'all 3');

    tree.insert(items[4]);
    expect(tree.data.height, equals(2), reason: 'height 4');
    expect(tree.all(), unorderedEquals(items), reason: 'all 4');
  });

  test('#insert forms a valid tree if items are inserted one by one', () {
    final tree = RBush(4);

    for (var i = 0; i < data.length; i++) {
      tree.insert(data[i]);
    }

    final tree2 = RBush(4).load(data);

    expect((tree.data.height - tree2.data.height).abs(), lessThanOrEqualTo(1));
    expect(tree.all(), unorderedEquals(tree2.all()));
  });

  test('#remove removes items correctly', () {
    final tree = RBush(4).load(data);
    final len = data.length;

    tree.remove(data[0]);
    tree.remove(data[1]);
    tree.remove(data[2]);

    tree.remove(data[len - 1]);
    tree.remove(data[len - 2]);
    tree.remove(data[len - 3]);

    expect(data.sublist(3, len - 3), unorderedEquals(tree.all()));
  });

  test('#remove does nothing if nothing found', () {
    final removed = RBush().load(data);
    removed.remove(listToBBox([13, 13, 13, 13]));

    expect(RBush().load(data).data, equals(removed.data));
  });

  test('#remove brings the tree to a clear state when removing everything one by one', () {
    final tree = RBush(4).load(data);

    for (final item in data) tree.remove(item);

    expect(tree.data, equals(RBush(4).data));
  });

  test('#clear should clear all the data in the tree', () {
    final tree = RBush(4).load(data);
    tree.clear();

    expect(tree.data, equals(RBush(4).data));
  });
}


test/rbush_knn.dart
import 'package:test/test.dart';
import 'package:rbush/rbush.dart';

RBushElement listToBBox(List<double> list) {
  return RBushElement(minX: list[0], minY: list[1], maxX: list[2], maxY: list[3], data: 0);
}

final data = <List<double>>[
  [87,55,87,56],[38,13,39,16],[7,47,8,47],[89,9,91,12],[4,58,5,60],[0,11,1,12],[0,5,0,6],[69,78,73,78],
  [56,77,57,81],[23,7,24,9],[68,24,70,26],[31,47,33,50],[11,13,14,15],[1,80,1,80],[72,90,72,91],[59,79,61,83],
  [98,77,101,77],[11,55,14,56],[98,4,100,6],[21,54,23,58],[44,74,48,74],[70,57,70,61],[32,9,33,12],[43,87,44,91],
  [38,60,38,60],[62,48,66,50],[16,87,19,91],[5,98,9,99],[9,89,10,90],[89,2,92,6],[41,95,45,98],[57,36,61,40],
  [50,1,52,1],[93,87,96,88],[29,42,33,42],[34,43,36,44],[41,64,42,65],[87,3,88,4],[56,50,56,52],[32,13,35,15],
  [3,8,5,11],[16,33,18,33],[35,39,38,40],[74,54,78,56],[92,87,95,90],[12,97,16,98],[76,39,78,40],[16,93,18,95],
  [62,40,64,42],[71,87,71,88],[60,85,63,86],[39,52,39,56],[15,18,19,18],[91,62,94,63],[10,16,10,18],[5,86,8,87],
  [85,85,88,86],[44,84,44,88],[3,94,3,97],[79,74,81,78],[21,63,24,66],[16,22,16,22],[68,97,72,97],[39,65,42,65],
  [51,68,52,69],[61,38,61,42],[31,65,31,65],[16,6,19,6],[66,39,66,41],[57,32,59,35],[54,80,58,84],[5,67,7,71],
  [49,96,51,98],[29,45,31,47],[31,72,33,74],[94,25,95,26],[14,7,18,8],[29,0,31,1],[48,38,48,40],[34,29,34,32],
  [99,21,100,25],[79,3,79,4],[87,1,87,5],[9,77,9,81],[23,25,25,29],[83,48,86,51],[79,94,79,95],[33,95,33,99],
  [1,14,1,14],[33,77,34,77],[94,56,98,59],[75,25,78,26],[17,73,20,74],[11,3,12,4],[45,12,47,12],[38,39,39,39],
  [99,3,103,5],[41,92,44,96],[79,40,79,41],[29,2,29,4],
].map((list) => listToBBox(list)).toList();

// Copied from `collections` package to reduce dependencies.
extension IndexedListExtension<E> on List<E> {
  /// Maps each element and its index to a new value.
  Iterable<R> mapIndexed<R>(R Function(int index, E element) convert) sync* {
    for (var index = 0; index < length; index++) {
      yield convert(index, this[index]);
    }
  }
}

void main() {
  test('finds n neighbours', () {
    final tree = RBush().load(data);
    final result = tree.knn(40, 40, 10);

    expect(result, equals(<List<double>>[
      [38,39,39,39],[35,39,38,40],[34,43,36,44],[29,42,33,42],[48,38,48,40],[31,47,33,50],[34,29,34,32],
      [29,45,31,47],[39,52,39,56],[57,36,61,40]
    ].map((list) => listToBBox(list)).toList()));
  });

  test('does not throw if requesting too many items', () {
    final tree = RBush().load(data);
    expect(() => tree.knn(40, 40, 1000), returnsNormally);
    List<RBushElement> result = [];
    try {
      result = tree.knn(40, 40, 1000);
    } on Error {};
    expect(result.length, equals(data.length));
  });

  test('finds all neighbors for maxDistance', () {
    final tree = RBush().load(data);
    final result = tree.knn(40, 40, 1, maxDistance: 10);

    expect(result, equals([listToBBox([38, 39, 39, 39])]));
  });

  test('does not throw if requesting too many items for maxDistance', () {
    final tree = RBush().load(data);
    expect(() => tree.knn(40, 40, 1000, maxDistance: 10), returnsNormally);
    List<RBushElement> result = [];
    try {
      result = tree.knn(40, 40, 1000, maxDistance: 10);
    } on Error {};
    expect(result, equals(<List<double>>[
      [38,39,39,39],[35,39,38,40],[34,43,36,44],[29,42,33,42],[48,38,48,40],[31,47,33,50],[34,29,34,32]
    ].map((list) => listToBBox(list)).toList()));
  });

  final pythData = <List<double>>[[0,0,0,0],[9,9,9,9],[12,12,12,12],[13,14,19,11]].map((list) => listToBBox(list));

  test('verify maxDistance excludes items too far away, in order to adhere to pythagoras theorem a^2+b^2=c^2', () {
    final tree = RBush().load(pythData);
    // sqrt(9^2+9^2)~=12.727
    final result = tree.knn(0, 0, 1000, maxDistance: 12.6);
    expect(result, equals([listToBBox([0, 0, 0, 0])]));
  });

  test('verify maxDistance includes all items within range, in order to adhere to pythagoras theorem a^2+b^2=c^2', () {
    final tree = RBush().load(pythData);
    // sqrt(9^2+9^2)~=12.727
    final result = tree.knn(0, 0, 1000, maxDistance: 12.8);
    expect(result, equals([listToBBox([0, 0, 0, 0]), listToBBox([9, 9, 9, 9])]));
  });

  final richData = <List<double>>[[1,2,1,2],[3,3,3,3],[5,5,5,5],[4,2,4,2],[2,4,2,4],[5,3,5,3]].mapIndexed((index, list) {
    return RBushElement(minX: list[0], minY: list[1], maxX: list[2], maxY: list[3], data: index + 1);
  });

  test('find n neighbours that do satisfy a given predicate', () {
    final tree = RBush().load(richData);
    final result = tree.knn(2, 4, 1, predicate: (item) => item.data < 5);
    expect(result.length, equals(1));

    final item = result.first;
    expect(item.minX, equals(3));
    expect(item.minY, equals(3));
    expect(item.maxX, equals(3));
    expect(item.maxY, equals(3));
    expect(item.data, equals(2));
  });
}


test/tinyqueue.dart
import 'package:rbush/rbush.dart' show TinyQueue;
import 'package:test/test.dart';
import 'dart:math' show Random;

void main() {
  final rand = Random();
  final data = [for (var i = 0; i < 100; i++) rand.nextInt(100)];
  final sorted = List.of(data);
  sorted.sort();

  test('maintains a priority queue', () {
    final queue = TinyQueue();
    for (final item in data) queue.push(item);

    expect(queue.peek(), equals(sorted.first));

    final result = <int>[];
    while (queue.isNotEmpty) result.add(queue.pop());

    expect(result, equals(sorted));
  });

  test('accepts data in constructor', () {
    final queue = TinyQueue(data);

    final result = <int>[];
    while (queue.isNotEmpty) result.add(queue.pop());

    expect(result, equals(sorted));
  });

  test('handles edge cases with few elements', () {
    final queue = TinyQueue();

    queue.push(2);
    queue.push(1);
    queue.pop();
    queue.pop();
    
    expect(() => queue.pop(), throwsStateError);

    queue.push(2);
    queue.push(1);

    expect(queue.pop(), equals(1));
    expect(queue.pop(), equals(2));
  });

  test('handles init with empty array', () {
    final queue = TinyQueue([]);
    expect(queue.data, equals([]));
  });
}


lib/src/tinyqueue.dart
// Copyright (c) 2021 Ilya Zverev, (c) 2019 Vladimir Agafonkin.
/// Port of https://github.com/mourner/tinyqueue.
// Use of this code is governed by an ISC license, see the LICENSE file.

/// The smallest and simplest binary heap priority queue.
///
/// ```dart
/// // create an empty priority queue
/// var queue = new TinyQueue();
///
/// // add some items
/// queue.push(7);
/// queue.push(5);
/// queue.push(10);
///
/// // remove the top item
/// var top = queue.pop(); // returns 5
///
/// // return the top item (without removal)
/// top = queue.peek(); // returns 7
///
/// // get queue length
/// queue.length; // returns 2
///
/// // create a priority queue from an existing array (modifies the array)
/// queue = new TinyQueue([7, 5, 10]);
///
/// // pass a custom item comparator as a second argument
/// queue = new TinyQueue([{value: 5}, {value: 7}], function (a, b) {
/// 	return a.value - b.value;
/// });
///
/// // turn a queue into a sorted array
/// var array = [];
/// while (queue.length) array.push(queue.pop());
/// ```
class TinyQueue<T> {
  final List<T> data;
  int length;
  final Comparator<T> compare;

  /// Constructs a new queue.
  ///
  /// Adding all elements at once would be faster than using [push]
  /// for each one. If elements are not `Comparable`, specify the
  /// [compare] function.
  TinyQueue([Iterable<T>? data, Comparator<T>? compare])
      : data = data?.toList() ?? [],
        length = data?.length ?? 0,
        compare = compare ?? ((a, b) => (a as Comparable).compareTo(b)) {
    if (length > 0) {
      for (var i = (length >> 1) - 1; i >= 0; i--) {
        _down(i);
      }
    }
  }

  bool get isEmpty => length == 0;
  bool get isNotEmpty => length > 0;

  /// Adds an element to this queue.
  push(T item) {
    data.add(item);
    _up(length++);
  }

  /// Removes the smallest element from the queue and returns it.
  T pop() {
    final top = data.first;
    final bottom = data.removeLast();

    if (--length > 0) {
      data[0] = bottom;
      _down(0);
    }

    return top;
  }

  /// Returns the smallest queue element without removing it.
  T peek() => data.first;

  _up(int pos) {
    final item = data[pos];
    while (pos > 0) {
      final parent = (pos - 1) >> 1;
      final current = data[parent];
      if (compare(item, current) >= 0) break;
      data[pos] = current;
      pos = parent;
    }
    data[pos] = item;
  }

  _down(int pos) {
    final halfLength = length >> 1;
    final item = data[pos];

    while (pos < halfLength) {
      int bestChild = (pos << 1) + 1; // initially it is the left child
      final right = bestChild + 1;

      if (right < length && compare(data[right], data[bestChild]) < 0) {
        bestChild = right;
      }
      if (compare(data[bestChild], item) >= 0) break;

      data[pos] = data[bestChild];
      pos = bestChild;
    }

    data[pos] = item;
  }
}


lib/src/quickselect.dart
// Copyright (c) 2021 Ilya Zverev, (c) 2018 Vladimir Agafonkin.
// Port of https://github.com/mourner/quickselect.
// Use of this code is governed by an ISC license, see the LICENSE file.

import 'dart:math' as math;

/// This function implements a fast
/// [selection algorithm](https://en.wikipedia.org/wiki/Selection_algorithm)
/// (specifically, [Floyd-Rivest selection](https://en.wikipedia.org/wiki/Floyd%E2%80%93Rivest_algorithm)).
///
/// Rearranges items so that all items in the `[left, k]` are the smallest.
/// The `k`-th element will have the `(k - left + 1)`-th smallest value in `[left, right]`.
///
/// - [arr]: the list to partially sort (in place)
/// - [k]: middle index for partial sorting (as defined above)
/// - [left]: left index of the range to sort (`0` by default)
/// - [right]: right index (last index of the array by default)
/// - [compare]: compare function, if items in the list are not `Comparable`.
///
/// Example:
///
/// ```dart
/// var arr = [65, 28, 59, 33, 21, 56, 22, 95, 50, 12, 90, 53, 28, 77, 39];
///
/// quickSelect(arr, 8);
///
/// // arr is [39, 28, 28, 33, 21, 12, 22, 50, 53, 56, 59, 65, 90, 77, 95]
/// //                                         ^^ middle index
/// ```
quickSelect<T>(List<T> arr, int k,
    [int left = 0, int? right, Comparator<T>? compare]) {
  if (arr.isEmpty) return;
  if (compare == null && arr.first is! Comparable) {
    throw ArgumentError(
        'Please either provide a comparator or use a list of Comparable elements.');
  }
  _quickSelectStep(arr, k, left, right ?? arr.length - 1,
      compare ?? (T a, T b) => (a as Comparable).compareTo(b));
}

_quickSelectStep<T>(
    List<T> arr, int k, int left, int right, Comparator<T> compare) {
  while (right > left) {
    if (right - left > 600) {
      final n = right - left + 1;
      final m = k - left + 1;
      final z = math.log(n);
      final s = 0.5 * math.exp(2 * z / 3);
      final sd = 0.5 * math.sqrt(z * s * (n - s) / n) * (m - n / 2 < 0 ? -1 : 1);
      final newLeft = math.max(left, (k - m * s / n + sd).floor());
      final newRight = math.min(right, (k + (n - m) * s / n + sd).floor());
      _quickSelectStep(arr, k, newLeft, newRight, compare);
    }

    final t = arr[k];
    var i = left;
    var j = right;

    _swap(arr, left, k);
    if (compare(arr[right], t) > 0) _swap(arr, left, right);

    while (i < j) {
      _swap(arr, i, j);
      i++;
      j--;
      while (compare(arr[i], t) < 0) i++;
      while (compare(arr[j], t) > 0) j--;
    }

    if (compare(arr[left], t) == 0) {
      _swap(arr, left, j);
    } else {
      j++;
      _swap(arr, j, right);
    }

    if (j <= k) left = j + 1;
    if (k <= j) right = j - 1;
  }
}

_swap<T>(List<T> arr, i, j) {
  final tmp = arr[i];
  arr[i] = arr[j];
  arr[j] = tmp;
}